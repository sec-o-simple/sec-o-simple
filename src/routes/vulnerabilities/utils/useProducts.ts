import { TProductTreeBranch } from '@/routes/products/types/tProductTreeBranch'
import useDocumentStore from '@/utils/useDocumentStore'
import { useMemo } from 'react'

export default function useProducts() {
  const productTree = useDocumentStore((state) => state.products)
  const productDetails = useMemo(
    () => Object.values(productTree)?.flatMap((ptb) => getProductDetails(ptb)),
    [productTree],
  )

  const getProductVersions = (productId: string) =>
    getProductVersionsOfBranch(Object.values(productTree), productId)

  return { productDetails, getProductVersions }
}

export type TProductDetails = {
  vendorName?: string
  productName: string
  productId: string
}

function getProductDetails(
  branch: TProductTreeBranch,
  vendorLabel?: string,
): TProductDetails[] {
  if (branch.category === 'vendor') {
    return branch.subBranches.flatMap((b) => getProductDetails(b, branch.name))
  } else if (branch.category === 'product_name') {
    return [
      {
        vendorName: vendorLabel,
        productName: branch.name,
        productId: branch.id,
      },
    ]
  } else {
    return []
  }
}

function getProductVersionsOfBranch(
  productTree: TProductTreeBranch[],
  productId: string,
): TProductTreeBranch[] | undefined {
  for (let branch of productTree) {
    if (branch.category === 'product_name' && branch.id === productId) {
      return branch.subBranches
    } else {
      const versions = getProductVersionsOfBranch(branch.subBranches, productId)
      if (versions) {
        return versions
      }
    }
  }
}
