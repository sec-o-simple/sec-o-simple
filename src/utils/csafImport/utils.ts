import { TProductTreeBranch } from '@/routes/products/types/tProductTreeBranch'

export function getParentPTB(
  id: string,
  startingBranches: TProductTreeBranch[],
  currentParent?: TProductTreeBranch,
): TProductTreeBranch | undefined {
  for (const branch of startingBranches) {
    if (startingBranches?.find((x) => x.id === id)) {
      return currentParent
    } else {
      const nestedSearch = getParentPTB(id, branch.subBranches, branch)
      if (nestedSearch) {
        return nestedSearch
      }
    }
  }
}
