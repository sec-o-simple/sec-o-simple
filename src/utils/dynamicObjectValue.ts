export type DynamicObjectValueKey<T, R = string> = keyof T | ((item: T) => R)

export function getDynamicObjectValue<T extends object, R = string>(
  obj: T,
  dynamicObjectValueKey: DynamicObjectValueKey<T, R>,
) {
  return typeof dynamicObjectValueKey === 'function'
    ? (dynamicObjectValueKey as (item: T) => R)(obj)
    : (obj[dynamicObjectValueKey] as R)
}
