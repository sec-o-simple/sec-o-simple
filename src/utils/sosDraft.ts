import { TDocumentInformation } from '@/routes/document-information/types/tDocumentInformation'
import {
  TProductFamily,
  TProductTreeBranch,
} from '@/routes/products/types/tProductTreeBranch'
import { TRelationship } from '@/routes/products/types/tRelationship'
import { TVulnerability } from '@/routes/vulnerabilities/types/tVulnerability'
import { getFilename } from './csafExport/helpers'
import { download } from './download'
import useDocumentStore, {
  TSOSDocumentType,
  sosDocumentTypes,
} from './useDocumentStore'

export type SOSDraft = {
  sosDocumentType: TSOSDocumentType
  documentInformation: TDocumentInformation
  products: TProductTreeBranch[]
  families: TProductFamily[]
  relationships: TRelationship[]
  vulnerabilities: TVulnerability[]
}

export function useSOSExport() {
  const documentStore = useDocumentStore()

  const exportSOSDocument = () => {
    download(
      `${getFilename(documentStore.documentInformation.id)}.sos.json`,
      JSON.stringify(documentStore, null, 2),
    )
  }

  return { exportSOSDocument }
}

export function useSOSImport() {
  const setSOSDocumentType = useDocumentStore(
    (state) => state.setSOSDocumentType,
  )
  const updateDocumentInformation = useDocumentStore(
    (state) => state.updateDocumentInformation,
  )
  const updateProducts = useDocumentStore((state) => state.updateProducts)
  const updateFamilies = useDocumentStore((state) => state.updateFamilies)
  const updateRelationships = useDocumentStore(
    (state) => state.updateRelationships,
  )
  const updateVulnerabilities = useDocumentStore(
    (state) => state.updateVulnerabilities,
  )

  const getSOSDraft = (jsonObject: object): SOSDraft | undefined => {
    if (
      !('sosDocumentType' in jsonObject) ||
      typeof jsonObject.sosDocumentType !== 'string' ||
      !sosDocumentTypes.includes(jsonObject.sosDocumentType as TSOSDocumentType)
    ) {
      return
    }

    if (
      !('documentInformation' in jsonObject) ||
      typeof jsonObject.documentInformation !== 'object'
    ) {
      return
    }

    if (
      !('products' in jsonObject) ||
      !(jsonObject.products instanceof Array)
    ) {
      return
    }

    if (
      !('families' in jsonObject) ||
      !(jsonObject.families instanceof Array)
    ) {
      return
    }

    if (
      !('relationships' in jsonObject) ||
      typeof jsonObject.relationships !== 'object'
    ) {
      return
    }

    if (
      !('vulnerabilities' in jsonObject) ||
      typeof jsonObject.vulnerabilities !== 'object'
    ) {
      return
    }

    return {
      sosDocumentType: jsonObject.sosDocumentType,
      documentInformation: jsonObject.documentInformation,
      products: jsonObject.products,
      families: jsonObject.families,
      relationships: jsonObject.relationships,
      vulnerabilities: jsonObject.vulnerabilities,
    } as SOSDraft
  }

  const isSOSDraft = (jsonObject: object): boolean => {
    return !!getSOSDraft(jsonObject)
  }

  const importSOSDocument = (jsonObject: object): boolean => {
    const sosDraft = getSOSDraft(jsonObject)
    if (sosDraft) {
      importSOSDraft(sosDraft)
      return true
    }
    return false
  }

  const importSOSDraft = (sosDraft: SOSDraft) => {
    setSOSDocumentType(sosDraft.sosDocumentType)
    updateDocumentInformation(sosDraft.documentInformation)
    updateProducts(sosDraft.products)
    updateFamilies(sosDraft.families)
    updateRelationships(sosDraft.relationships)
    updateVulnerabilities(sosDraft.vulnerabilities)
  }

  return { isSOSDraft, importSOSDocument, importSOSDraft }
}
