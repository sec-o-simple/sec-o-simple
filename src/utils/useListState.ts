import { useState } from 'react'
import {
  DynamicObjectValueKey,
  getDynamicObjectValue,
} from './dynamicObjectValue'

export type UseListStateProps<T, IdType> = {
  initialData?: T[]
  idKey?: DynamicObjectValueKey<T, IdType>
  /** Generator function to create a new element */
  generator?: T | (() => T)
  onRemove?: (entry: T) => void
}

export type ListState<T, IdType = string> = {
  data: T[]
  setData: React.Dispatch<React.SetStateAction<T[]>>
  updateDataEntry: (updatedEntry: T) => void
  removeDataEntry: (entry: T) => void
  addDataEntry: () => T | undefined
  getId: (entry: T) => IdType
}

export function useListState<T extends object, IdType = string>(
  options?: UseListStateProps<T, IdType>,
): ListState<T, IdType> {
  const { initialData = [], idKey = 'id' as keyof T, generator } = options ?? {}
  const [data, setData] = useState<T[]>(initialData)

  const updateDataEntry = (updatedEntry: T) => {
    setData(
      data.map((e) => (getId(e) === getId(updatedEntry) ? updatedEntry : e)),
    )
  }

  const removeDataEntry = (entry: T) => {
    setData([...data.filter((e) => getId(e) !== getId(entry))])
    options?.onRemove?.(entry)
  }

  /** adds a new data entry to the state
   *  returns the id of the new data entry */
  const addDataEntry = () => {
    const entry = typeof generator === 'function' ? generator() : generator
    if (entry) {
      setData([...data, entry])
      return entry
    }
  }

  const getId = (entry: T) => getDynamicObjectValue(entry, idKey)

  return {
    data,
    setData,
    updateDataEntry,
    removeDataEntry,
    getId,
    addDataEntry,
  }
}
