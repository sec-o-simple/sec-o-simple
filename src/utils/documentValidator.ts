import { TDocumentStore } from '@/utils/useDocumentStore'
import { createCSAFDocument } from './csafExport/csafExport'

import validate from '@secvisogram/csaf-validator-lib/validate.js'
import * as basic from '@secvisogram/csaf-validator-lib/basic.js'
import { ValidationMessage } from './useValidationStore'

const tests: unknown[] = [...Object.values(basic)]

export interface ValidationResult {
  isValid: boolean
  messages: ValidationMessage[]
}

export async function validateDocument(
  documentStore: TDocumentStore,
): Promise<ValidationResult> {
  try {
    const csafDocument = createCSAFDocument(documentStore)
    const result = await validate(tests, csafDocument)

    // Collect all messages with their severity
    const messages: ValidationMessage[] = []

    result.tests.forEach((test) => {
      test.errors.forEach((error) => {
        messages.push({
          path: error.instancePath,
          message: error.message,
          severity: 'error',
        })
      })

      test.warnings.forEach((warning) => {
        messages.push({
          path: warning.instancePath,
          message: warning.message,
          severity: 'warning',
        })
      })

      test.infos.forEach((info) => {
        messages.push({
          path: info.instancePath,
          message: info.message,
          severity: 'info',
        })
      })
    })

    return {
      isValid: result.isValid,
      messages: messages,
    }
  } catch (error) {
    return {
      isValid: false,
      messages: [],
    }
  }
}
