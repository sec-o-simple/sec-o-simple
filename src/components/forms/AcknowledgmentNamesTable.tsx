import { Input } from '@/components/forms/Input'
import {
  getDefaultDocumentAcknowledgmentName,
  TAcknowledgment,
  TAcknowledgmentName,
} from '@/routes/document-information/types/tDocumentAcknowledgments'
import { useListState } from '@/utils/useListState'
import { faAdd, faTrash } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Button } from '@heroui/button'
import {
  Table,
  TableBody,
  TableCell,
  TableColumn,
  TableHeader,
  TableRow,
} from '@heroui/table'
import { useEffect } from 'react'
import { useTranslation } from 'react-i18next'
import IconButton from './IconButton'

export default function AcknowledgmentNamesTable({
  acknowledgment,
  acknowledgmentIndex,
  onChange,
}: {
  acknowledgment: TAcknowledgment
  acknowledgmentIndex: number
  onChange: (acknowledgment: TAcknowledgment) => void
}) {
  const { t } = useTranslation()
  const acknowledgmentNames = useListState<TAcknowledgmentName>({
    initialData: acknowledgment.names,
    generator: getDefaultDocumentAcknowledgmentName,
  })

  useEffect(
    () =>
      onChange({
        ...acknowledgment,
        names: acknowledgmentNames.data,
      }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [acknowledgmentNames.data],
  )

  const handleAddRevision = () => acknowledgmentNames.addDataEntry()
  const handleRevisionChange = (updatedName: TAcknowledgmentName) =>
    acknowledgmentNames.updateDataEntry(updatedName)
  const handleDeleteName = (name: TAcknowledgmentName) =>
    acknowledgmentNames.removeDataEntry(name)

  return (
    <div className="mt-4">
      <div className="mb-4 flex items-center justify-between">
        <h2 className="text-lg font-semibold">
          {t('document.acknowledgments.names')}
        </h2>
        <Button variant="light" color="primary" onPress={handleAddRevision}>
          <FontAwesomeIcon icon={faAdd} />
          {t('common.add', {
            label: t('document.acknowledgments.name'),
          })}
        </Button>
      </div>

      <Table>
        <TableHeader>
          <TableColumn width="20%">
            {t('document.acknowledgments.name')}
          </TableColumn>
          <TableColumn width="10%">{t('common.actions')}</TableColumn>
        </TableHeader>
        <TableBody emptyContent={t('document.acknowledgments.empty')}>
          {acknowledgmentNames.data?.map((name, index) => (
            <TableRow key={name.name}>
              <TableCell>
                <Input
                  value={name.name}
                  csafPath={`/document/acknowledgments/${acknowledgmentIndex}/names/${index}`}
                  onValueChange={(value) =>
                    handleRevisionChange({
                      ...acknowledgmentNames.data[index],
                      name: value,
                    })
                  }
                  placeholder={t('common.placeholder', {
                    label: t('document.acknowledgments.name'),
                  })}
                />
              </TableCell>
              <TableCell>
                <IconButton
                  icon={faTrash}
                  onPress={() => handleDeleteName(name)}
                />
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </div>
  )
}
