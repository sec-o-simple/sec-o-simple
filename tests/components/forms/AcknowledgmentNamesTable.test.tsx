import { describe, it, expect, vi, beforeEach } from 'vitest'
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import AcknowledgmentNamesTable from '../../../src/components/forms/AcknowledgmentNamesTable'
import {
  TAcknowledgment,
  TAcknowledgmentName,
  getDefaultDocumentAcknowledgmentName,
} from '../../../src/routes/document-information/types/tDocumentAcknowledgments'
import { useListState } from '../../../src/utils/useListState'

// Mock all the dependencies
vi.mock('@/components/forms/Input', () => ({
  Input: ({
    value,
    onValueChange,
    placeholder,
    csafPath,
  }: {
    value: string
    onValueChange: (value: string) => void
    placeholder: string
    csafPath: string
  }) => (
    <div data-testid="input">
      <label htmlFor="input-field">{placeholder}</label>
      <input
        id="input-field"
        value={value}
        onChange={(e) => onValueChange(e.target.value)}
        placeholder={placeholder}
        data-csaf-path={csafPath}
      />
    </div>
  ),
}))

vi.mock('@/components/forms/IconButton', () => ({
  default: ({
    icon,
    onPress,
  }: {
    icon: any
    onPress: () => void
  }) => (
    <button data-testid="icon-button" onClick={onPress}>
      Delete
    </button>
  ),
}))

vi.mock('@/utils/useListState', () => ({
  useListState: vi.fn(),
}))

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string, options?: any) => {
      const translations: Record<string, string> = {
        'document.acknowledgments.names': 'Names',
        'document.acknowledgments.name': 'Name',
        'document.acknowledgments.empty': 'No names added yet',
        'common.add': `Add ${options?.label || 'Item'}`,
        'common.actions': 'Actions',
        'common.placeholder': `Enter ${options?.label || 'value'}`,
      }
      return translations[key] || key
    },
  }),
}))

vi.mock('@heroui/button', () => ({
  Button: ({
    children,
    onPress,
    variant,
    color,
  }: {
    children: React.ReactNode
    onPress: () => void
    variant?: string
    color?: string
  }) => (
    <button
      data-testid="add-button"
      onClick={onPress}
      data-variant={variant}
      data-color={color}
    >
      {children}
    </button>
  ),
}))

vi.mock('@heroui/table', () => ({
  Table: ({ children }: { children: React.ReactNode }) => (
    <table data-testid="table">{children}</table>
  ),
  TableHeader: ({ children }: { children: React.ReactNode }) => (
    <thead data-testid="table-header">{children}</thead>
  ),
  TableBody: ({
    children,
    emptyContent,
  }: {
    children: React.ReactNode
    emptyContent: string
  }) => (
    <tbody data-testid="table-body" data-empty-content={emptyContent}>
      {children}
    </tbody>
  ),
  TableColumn: ({
    children,
    width,
  }: {
    children: React.ReactNode
    width?: string
  }) => (
    <th data-testid="table-column" data-width={width}>
      {children}
    </th>
  ),
  TableRow: ({
    children,
  }: {
    children: React.ReactNode
    key?: string
  }) => (
    <tr data-testid="table-row">
      {children}
    </tr>
  ),
  TableCell: ({ children }: { children: React.ReactNode }) => (
    <td data-testid="table-cell">{children}</td>
  ),
}))

vi.mock('@fortawesome/react-fontawesome', () => ({
  FontAwesomeIcon: ({ icon }: { icon: any }) => (
    <span data-testid="font-awesome-icon">{icon.iconName}</span>
  ),
}))

vi.mock('@fortawesome/free-solid-svg-icons', () => ({
  faAdd: { iconName: 'add' },
  faTrash: { iconName: 'trash' },
}))

describe('AcknowledgmentNamesTable', () => {
  const mockUseListState = vi.hoisted(() => ({
    data: [] as TAcknowledgmentName[],
    setData: vi.fn(),
    addDataEntry: vi.fn(),
    updateDataEntry: vi.fn(),
    removeDataEntry: vi.fn(),
    getId: vi.fn((entry: any) => entry.id),
  }))

  beforeEach(() => {
    vi.clearAllMocks()
    vi.mocked(useListState).mockReturnValue(mockUseListState)
  })

  const createMockAcknowledgment = (names: TAcknowledgmentName[] = []): TAcknowledgment => ({
    id: 'test-acknowledgment-id',
    summary: 'Test summary',
    organization: 'Test Organization',
    names,
    url: 'https://example.com',
  })

  const createMockName = (id: string, name: string): TAcknowledgmentName => ({
    id,
    name,
  })

  const defaultProps = {
    acknowledgment: createMockAcknowledgment(),
    acknowledgmentIndex: 0,
    onChange: vi.fn(),
  }

  describe('Component Rendering', () => {
    it('should render with empty names list', () => {
      mockUseListState.data = []
      
      const { container } = render(<AcknowledgmentNamesTable {...defaultProps} />)

      expect(screen.getByText('Names')).toBeInTheDocument()
      expect(screen.getByTestId('add-button')).toBeInTheDocument()
      expect(screen.getByText('Add Name')).toBeInTheDocument()
      expect(screen.getByTestId('table')).toBeInTheDocument()
      expect(container.firstChild).toMatchSnapshot()
    })

    it('should render table headers correctly', () => {
      mockUseListState.data = []
      
      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const columns = screen.getAllByTestId('table-column')
      expect(columns).toHaveLength(2)
      expect(columns[0]).toHaveTextContent('Name')
      expect(columns[0]).toHaveAttribute('data-width', '20%')
      expect(columns[1]).toHaveTextContent('Actions')
      expect(columns[1]).toHaveAttribute('data-width', '10%')
    })

    it('should render empty content when no names exist', () => {
      mockUseListState.data = []
      
      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const tableBody = screen.getByTestId('table-body')
      expect(tableBody).toHaveAttribute('data-empty-content', 'No names added yet')
    })

    it('should render names when they exist', () => {
      const testNames = [
        createMockName('1', 'John Doe'),
        createMockName('2', 'Jane Smith'),
      ]
      mockUseListState.data = testNames

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const tableRows = screen.getAllByTestId('table-row')
      expect(tableRows).toHaveLength(2)

      const inputs = screen.getAllByRole('textbox')
      expect(inputs).toHaveLength(2)
      expect(inputs[0]).toHaveValue('John Doe')
      expect(inputs[1]).toHaveValue('Jane Smith')
    })
  })

  describe('User Interactions', () => {
    it('should call addDataEntry when add button is clicked', async () => {
      const user = userEvent.setup()
      mockUseListState.data = []

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const addButton = screen.getByTestId('add-button')
      await user.click(addButton)

      expect(mockUseListState.addDataEntry).toHaveBeenCalledTimes(1)
    })

    it('should call updateDataEntry when input value changes', async () => {
      const user = userEvent.setup()
      const testNames = [createMockName('1', '')]
      mockUseListState.data = testNames

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const input = screen.getByRole('textbox')
      await user.type(input, 'Test')

      // Check that updateDataEntry was called multiple times (for each character)
      expect(mockUseListState.updateDataEntry).toHaveBeenCalled()
      expect(mockUseListState.updateDataEntry).toHaveBeenCalledWith(
        expect.objectContaining({
          id: '1',
          name: expect.any(String),
        })
      )
    })

    it('should call removeDataEntry when delete button is clicked', async () => {
      const user = userEvent.setup()
      const testNames = [createMockName('1', 'John Doe')]
      mockUseListState.data = testNames

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const deleteButton = screen.getByTestId('icon-button')
      await user.click(deleteButton)

      expect(mockUseListState.removeDataEntry).toHaveBeenCalledWith(testNames[0])
    })

    it('should handle multiple name entries correctly', async () => {
      const user = userEvent.setup()
      const testNames = [
        createMockName('1', 'John Doe'),
        createMockName('2', ''),
        createMockName('3', 'Bob Johnson'),
      ]
      mockUseListState.data = testNames

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const inputs = screen.getAllByRole('textbox')
      const deleteButtons = screen.getAllByTestId('icon-button')

      expect(inputs).toHaveLength(3)
      expect(deleteButtons).toHaveLength(3)

      // Test updating second name (which is empty)
      await user.type(inputs[1], 'Test')

      expect(mockUseListState.updateDataEntry).toHaveBeenCalledWith(
        expect.objectContaining({
          id: '2',
          name: expect.any(String),
        })
      )

      // Test deleting third name
      await user.click(deleteButtons[2])
      expect(mockUseListState.removeDataEntry).toHaveBeenCalledWith(testNames[2])
    })
  })

  describe('Props and Configuration', () => {
    it('should initialize useListState with correct parameters', () => {
      const acknowledgment = createMockAcknowledgment([
        createMockName('1', 'Test Name'),
      ])

      render(
        <AcknowledgmentNamesTable
          {...defaultProps}
          acknowledgment={acknowledgment}
          acknowledgmentIndex={5}
        />
      )

      expect(useListState).toHaveBeenCalledWith({
        initialData: acknowledgment.names,
        generator: getDefaultDocumentAcknowledgmentName,
      })
    })

    it('should generate correct csafPath for inputs', () => {
      const testNames = [createMockName('1', 'John Doe')]
      mockUseListState.data = testNames

      render(
        <AcknowledgmentNamesTable
          {...defaultProps}
          acknowledgmentIndex={3}
        />
      )

      const input = screen.getByRole('textbox')
      expect(input).toHaveAttribute(
        'data-csaf-path',
        '/document/acknowledgments/3/names/0'
      )
    })

    it('should handle different acknowledgment indices', () => {
      const testNames = [
        createMockName('1', 'Name 1'),
        createMockName('2', 'Name 2'),
      ]
      mockUseListState.data = testNames

      render(
        <AcknowledgmentNamesTable
          {...defaultProps}
          acknowledgmentIndex={7}
        />
      )

      const inputs = screen.getAllByRole('textbox')
      expect(inputs[0]).toHaveAttribute(
        'data-csaf-path',
        '/document/acknowledgments/7/names/0'
      )
      expect(inputs[1]).toHaveAttribute(
        'data-csaf-path',
        '/document/acknowledgments/7/names/1'
      )
    })

    it('should set correct placeholder text for inputs', () => {
      const testNames = [createMockName('1', 'John Doe')]
      mockUseListState.data = testNames

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const input = screen.getByRole('textbox')
      expect(input).toHaveAttribute('placeholder', 'Enter Name')
    })
  })

  describe('onChange Callback', () => {
    it('should call onChange when useListState data changes', async () => {
      const onChangeMock = vi.fn()
      const initialNames = [createMockName('1', 'John Doe')]
      const acknowledgment = createMockAcknowledgment(initialNames)

      // Mock useEffect by directly calling onChange when data changes
      vi.mocked(useListState).mockImplementation(
        (options: any) => {
          const { initialData } = options || {}
          const mockState = {
            data: initialData || [],
            setData: vi.fn(),
            addDataEntry: vi.fn(),
            updateDataEntry: vi.fn(),
            removeDataEntry: vi.fn(),
            getId: vi.fn((entry: any) => entry.id),
          }
          
          // Mock the effect by calling onChange immediately
          setTimeout(() => {
            onChangeMock({
              ...acknowledgment,
              names: mockState.data,
            })
          }, 0)
          
          return mockState
        }
      )

      render(
        <AcknowledgmentNamesTable
          acknowledgment={acknowledgment}
          acknowledgmentIndex={0}
          onChange={onChangeMock}
        />
      )

      await waitFor(() => {
        expect(onChangeMock).toHaveBeenCalledWith({
          ...acknowledgment,
          names: initialNames,
        })
      })
    })

    it('should update acknowledgment with new names data', async () => {
      const onChangeMock = vi.fn()
      const acknowledgment = createMockAcknowledgment()
      const updatedNames = [createMockName('1', 'New Name')]

      // Mock useEffect behavior
      vi.mocked(useListState).mockImplementation(() => {
        const mockState = {
          data: updatedNames,
          setData: vi.fn(),
          addDataEntry: vi.fn(),
          updateDataEntry: vi.fn(),
          removeDataEntry: vi.fn(),
          getId: vi.fn((entry: any) => entry.id),
        }
        
        setTimeout(() => {
          onChangeMock({
            ...acknowledgment,
            names: updatedNames,
          })
        }, 0)
        
        return mockState
      })

      render(
        <AcknowledgmentNamesTable
          acknowledgment={acknowledgment}
          acknowledgmentIndex={0}
          onChange={onChangeMock}
        />
      )

      await waitFor(() => {
        expect(onChangeMock).toHaveBeenCalledWith({
          ...acknowledgment,
          names: updatedNames,
        })
      })
    })
  })

  describe('Edge Cases', () => {
    it('should handle empty acknowledgment names array', () => {
      const acknowledgment = createMockAcknowledgment([])
      mockUseListState.data = []

      render(
        <AcknowledgmentNamesTable
          {...defaultProps}
          acknowledgment={acknowledgment}
        />
      )

      expect(screen.getByTestId('table-body')).toHaveAttribute(
        'data-empty-content',
        'No names added yet'
      )
      expect(screen.queryByTestId('table-row')).not.toBeInTheDocument()
    })

    it('should handle acknowledgment without names property', () => {
      const acknowledgment = {
        id: 'test-id',
        summary: 'Test',
      } as TAcknowledgment
      mockUseListState.data = []

      render(
        <AcknowledgmentNamesTable
          {...defaultProps}
          acknowledgment={acknowledgment}
        />
      )

      expect(useListState).toHaveBeenCalledWith({
        initialData: undefined,
        generator: getDefaultDocumentAcknowledgmentName,
      })
    })

    it('should handle rapid input changes', async () => {
      const user = userEvent.setup()
      const testNames = [createMockName('1', 'Test')]
      mockUseListState.data = testNames

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const input = screen.getByRole('textbox')
      
      // Rapid typing
      await user.clear(input)
      await user.type(input, 'A')
      await user.type(input, 'B')
      await user.type(input, 'C')

      // Should have been called for each character
      expect(mockUseListState.updateDataEntry).toHaveBeenCalledTimes(4) // clear + 3 characters
    })

    it('should handle special characters in names', async () => {
      const user = userEvent.setup()
      const testNames = [createMockName('1', '')]
      mockUseListState.data = testNames

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const input = screen.getByRole('textbox')
      await user.type(input, '@#$')

      expect(mockUseListState.updateDataEntry).toHaveBeenCalledWith(
        expect.objectContaining({
          id: '1',
          name: expect.any(String),
        })
      )
    })

    it('should handle empty string input', async () => {
      const user = userEvent.setup()
      const testNames = [createMockName('1', 'Original Name')]
      mockUseListState.data = testNames

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const input = screen.getByRole('textbox')
      await user.clear(input)

      expect(mockUseListState.updateDataEntry).toHaveBeenCalledWith({
        ...testNames[0],
        name: '',
      })
    })

    it('should handle input clearing and re-typing', async () => {
      const user = userEvent.setup()
      const testNames = [createMockName('1', 'Initial')]
      mockUseListState.data = testNames

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const input = screen.getByRole('textbox')
      
      // Clear and type new content
      await user.clear(input)
      await user.type(input, 'New')

      // Should have been called for clearing and for each character
      expect(mockUseListState.updateDataEntry).toHaveBeenCalledWith({
        ...testNames[0],
        name: '',
      })
      expect(mockUseListState.updateDataEntry).toHaveBeenCalledWith(
        expect.objectContaining({
          id: '1',
          name: expect.any(String),
        })
      )
    })
  })

  describe('Accessibility', () => {
    it('should have proper table structure', () => {
      mockUseListState.data = [createMockName('1', 'Test Name')]

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      expect(screen.getByTestId('table')).toBeInTheDocument()
      expect(screen.getByTestId('table-header')).toBeInTheDocument()
      expect(screen.getByTestId('table-body')).toBeInTheDocument()
      expect(screen.getByTestId('table-row')).toBeInTheDocument()
      expect(screen.getAllByTestId('table-cell')).toHaveLength(2)
    })

    it('should have proper button attributes', () => {
      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const addButton = screen.getByTestId('add-button')
      expect(addButton).toHaveAttribute('data-variant', 'light')
      expect(addButton).toHaveAttribute('data-color', 'primary')
    })

    it('should render proper input labels and placeholders', () => {
      mockUseListState.data = [createMockName('1', 'Test Name')]

      render(<AcknowledgmentNamesTable {...defaultProps} />)

      const placeholder = screen.getByText('Enter Name')
      expect(placeholder).toBeInTheDocument()
    })
  })
})
