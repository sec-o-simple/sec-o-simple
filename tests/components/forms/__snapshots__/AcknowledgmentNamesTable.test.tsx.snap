// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AcknowledgmentNamesTable > Component Rendering > should render with empty names list 1`] = `
<div
  class="mt-4"
>
  <div
    class="mb-4 flex items-center justify-between"
  >
    <h2
      class="text-lg font-semibold"
    >
      Names
    </h2>
    <button
      data-color="primary"
      data-testid="add-button"
      data-variant="light"
    >
      <span
        data-testid="font-awesome-icon"
      >
        add
      </span>
      Add Name
    </button>
  </div>
  <table
    data-testid="table"
  >
    <thead
      data-testid="table-header"
    >
      <th
        data-testid="table-column"
        data-width="20%"
      >
        Name
      </th>
      <th
        data-testid="table-column"
        data-width="10%"
      >
        Actions
      </th>
    </thead>
    <tbody
      data-empty-content="No names added yet"
      data-testid="table-body"
    />
  </table>
</div>
`;
